### PR TITLE
NMA-331 DashPay: Notifications Contact Requests

### PR DESCRIPTION
**Pull request by HashEngineering**
## Issue being fixed or feature implemented

1. Notifications Screen (accessible from the Home Screen Bell or Number icon by avatar)

### CODE REVIEW

File: wallet/src/de/schildbach/wallet/ui/HeaderBalanceFragment.java [hunk 1]
```diff
@@ -18,33 +18,38 @@
...
 import androidx.lifecycle.Observer;
+import androidx.lifecycle.ViewModelProvider;
 import androidx.lifecycle.ViewModelProviders;
 import androidx.loader.app.LoaderManager;
 import androidx.loader.content.Loader;
 
 import com.amulyakhare.textdrawable.TextDrawable;
 
 import org.bitcoinj.core.Coin;
+import org.bitcoinj.core.Utils;
```
Comment (sambarboza) [inline wallet/src/de/schildbach/wallet/ui/HeaderBalanceFragment.java:44]: It looks like these imports aren't being used.

File: wallet/src/de/schildbach/wallet/ui/dashpay/ContactSearchResultsAdapter.kt [hunk 2]
```diff
@@ -225,16 +225,18 @@ class ContactSearchResultsAdapter(private val onSortOrderChangedListener: OnSort
 
                 view_all_contacts.setOnClickListener {
                     //TODO: Show View All Contact Request Activity
```
Comment (sambarboza) [inline wallet/src/de/schildbach/wallet/ui/dashpay/ContactSearchResultsAdapter.kt:227]: Can we remove this //TODO?

File: wallet/src/de/schildbach/wallet/ui/dashpay/ContactsActivity.kt [hunk 3]
```diff
@@ -220,4 +238,8 @@ class ContactsActivity : GlobalFooterActivity(), TextWatcher, ContactSearchResul
         }
         return super.onOptionsItemSelected(item)
     }
+
+    override fun onViewAllRequests() {
+        startActivity(ContactsActivity.createIntent(this, ContactsActivity.MODE_VIEW_REQUESTS))
```
Comment (sambarboza) [inline wallet/src/de/schildbach/wallet/ui/dashpay/ContactsActivity.kt:243]: `ContactsActivity` can be removed from this line in the `.createIntent` call and the `.MODE_REVIEW_REQUESTS` access.

File: wallet/src/de/schildbach/wallet/ui/dashpay/NotificationsActivity.kt [hunk 4]
```diff
@@ -0,0 +1,229 @@
+package de.schildbach.wallet.ui.dashpay
```
Comment (sambarboza) [inline wallet/src/de/schildbach/wallet/ui/dashpay/NotificationsActivity.kt:1]: Please add license header.

File: wallet/src/de/schildbach/wallet/ui/dashpay/NotificationsAdapter.kt [hunk 5]
```diff
@@ -0,0 +1,259 @@
+package de.schildbach.wallet.ui.dashpay
```
Comment (sambarboza) [inline wallet/src/de/schildbach/wallet/ui/dashpay/NotificationsAdapter.kt:1]: Please add license header.

File: wallet/src/de/schildbach/wallet/ui/dashpay/NotificationsAdapter.kt [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+import android.widget.ImageView
+import android.widget.TextView
+import androidx.recyclerview.widget.RecyclerView
+import com.bumptech.glide.Glide
+import de.schildbach.wallet.data.DashPayContactRequest
+import de.schildbach.wallet.data.UsernameSearchResult
+import de.schildbach.wallet.ui.UserAvatarPlaceholderDrawable
+import de.schildbach.wallet_test.R
+import kotlinx.android.synthetic.main.contact_request_row.view.*
+import org.bitcoinj.core.Utils
```
Comment (sambarboza) [inline wallet/src/de/schildbach/wallet/ui/dashpay/NotificationsAdapter.kt:16]: It looks like this import can be removed.

File: wallet/src/de/schildbach/wallet/ui/dashpay/PlatformRepo.kt [hunk 7]
```diff
@@ -50,7 +51,12 @@ import org.dashevo.dpp.util.HashUtils
 import org.dashevo.platform.Names
 import org.dashevo.platform.Platform
 import org.slf4j.LoggerFactory
+import java.util.*
 import java.util.concurrent.TimeoutException
+import kotlin.collections.ArrayList
+import kotlin.collections.HashMap
+import kotlin.collections.HashSet
+import kotlin.math.max
```
[the remark this answers is not in the document]
Comment (HashEngineering) [reply wallet/src/de/schildbach/wallet/ui/dashpay/PlatformRepo.kt:59]: > The bottom margin from the date needs to be ajusted. Also, the date formatting is different from layout, I wonder which one should be considered correct. The date formatting seems to be the same used on the transaction history, so I would keep this same one as well. The texts seems to be different as well.
> 
> Screen Shot 2020-06-30 at 2 57 32 PM
> 
> 

Here is the updated layout.  Turns out that the height of an accepted contact request is 79dp, while contact requests are 70dp.

new-format


The text should differ based on who accepts the request:

If they sent it, then you accepted, then "You have accepted the contact request from username"

If you sent it, then they accept it, then "Username has accepted your contact request"

For example (without the corrected layout):
Screenshot_1593557306

[the remark this answers is not in the document]
Comment (HashEngineering) [reply wallet/src/de/schildbach/wallet/ui/dashpay/PlatformRepo.kt]: unused imports removed

**Comment by HashEngineering** [comment]
> The bottom margin from the date needs to be ajusted. Also, the date formatting is different from layout, I wonder which one should be considered correct. The date formatting seems to be the same used on the transaction history, so I would keep this same one as well. The texts seems to be different as well.
> 
> Screen Shot 2020-06-30 at 2 57 32 PM
> 
> 

Here is the updated layout.  Turns out that the height of an accepted contact request is 79dp, while contact requests are 70dp.

new-format


The text should differ based on who accepts the request:

If they sent it, then you accepted, then "You have accepted the contact request from username"

If you sent it, then they accept it, then "Username has accepted your contact request"

For example (without the corrected layout):
Screenshot_1593557306

File: wallet/src/de/schildbach/wallet/ui/HeaderBalanceFragment.java (hunk 1 above)
Comment (HashEngineering) [reply wallet/src/de/schildbach/wallet/ui/HeaderBalanceFragment.java]: unused imports removed

File: wallet/src/de/schildbach/wallet/ui/dashpay/ContactsActivity.kt (hunk 3 above)
Comment (HashEngineering) [reply wallet/src/de/schildbach/wallet/ui/dashpay/ContactsActivity.kt]: yes, and its removed

File: wallet/src/de/schildbach/wallet/ui/dashpay/NotificationsActivity.kt (hunk 4 above)
Comment (HashEngineering) [reply wallet/src/de/schildbach/wallet/ui/dashpay/NotificationsActivity.kt:17]: added

File: wallet/src/de/schildbach/wallet/ui/dashpay/NotificationsAdapter.kt (hunk 5 above)
Comment (HashEngineering) [reply wallet/src/de/schildbach/wallet/ui/dashpay/NotificationsAdapter.kt:17]: added

File: wallet/src/de/schildbach/wallet/ui/dashpay/NotificationsAdapter.kt (hunk 6 above)
Comment (HashEngineering) [reply wallet/src/de/schildbach/wallet/ui/dashpay/NotificationsAdapter.kt]: removed

File: wallet/src/de/schildbach/wallet/ui/dashpay/ContactSearchResultsAdapter.kt (hunk 2 above)
Comment (HashEngineering) [reply wallet/src/de/schildbach/wallet/ui/dashpay/ContactSearchResultsAdapter.kt]: removed

**Pull request merged**